File: backend/app/contracts/run_budget_contract.py
"""
Run-Budget-Contract v1 (Pydantic v2, Issue #764).

Kanonischer, versionierter Vertrag für Kosten-, Token- und Zeitbudgets:
  - RunBudgetConfig: vom Nutzer gesetzte Limits (weich/hart)
  - BudgetWarning: auditierbare Warnung bei Limitüberschreitung
  - UsageMetrics: gemessener Verbrauch (Tokens, Kosten, Laufzeit, Aufrufe)
  - RunUsage: Verbrauch gesamt + pro Stage/Provider/Modell
  - RunBudgetStatus: Config + aktueller Verbrauch + Warnungen
  - PreflightEstimate: ehrlich gekennzeichnete Schätzung vor Run-Start

Regeln:
  - Geldbeträge ausschließlich als Integer-Micros (1 Einheit = 10^-6 Währung),
    niemals Floats.
  - Fehlende Providerdaten werden niemals als 0 ausgegeben: optionale Felder
    bleiben None, der Status kennzeichnet unknown/estimated/free/measured.
  - Lokale Modelle ohne Geldpreis: cost_status="free".
"""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

_STRICT = ConfigDict(extra="forbid")

BudgetDimension = Literal["tokens", "cost", "time", "calls"]
BudgetEnforcement = Literal["soft", "hard"]
CostStatus = Literal["measured", "estimated", "free", "unknown"]
TokensStatus = Literal["measured", "partial", "unknown"]
MeasurementStatus = Literal["complete", "partial", "unknown"]
DataQuality = Literal["high", "medium", "low", "unknown"]
BudgetState = Literal["ok", "warning", "exceeded"]

# Abbruchgründe auf Run-Ebene (Issue #764): Budgetabbruch muss von
# technischem Fehler und Nutzerabbruch unterscheidbar sein.
TerminationReason = Literal[
    "completed",
    "error",
    "user_cancel",
    "user_stop",
    "budget_tokens",
    "budget_cost",
    "budget_time",
    "budget_calls",
]


class RunBudgetConfig(BaseModel):
    """Vom Nutzer gesetzte Budget-Limits für einen Run.

    Alle Limits optional; None bedeutet „kein Limit". `enforcement` gilt
    für alle gesetzten Limits gemeinsam: soft = warnen und weiterlaufen,
    hard = weitere planbare Modellaufrufe deterministisch verhindern.
    """

    model_config = _STRICT

    schema_version: Literal[1] = 1
    max_tokens: Optional[int] = Field(None, ge=1)
    max_cost_micros: Optional[int] = Field(None, ge=1)
    max_duration_seconds: Optional[int] = Field(None, ge=1)
    max_llm_calls: Optional[int] = Field(None, ge=1)
    enforcement: BudgetEnforcement = "soft"
    currency: str = Field("USD", min_length=3, max_length=3)


class BudgetWarning(BaseModel):
    """Auditierbare Warnung bei (weicher oder harter) Limitüberschreitung."""

    model_config = _STRICT

    dimension: BudgetDimension
    severity: BudgetEnforcement
    threshold: int = Field(ge=0)
    observed: int = Field(ge=0)
    message: str = Field(min_length=1)
    ts: str = Field(min_length=1)  # ISO-8601 UTC


class UsageMetrics(BaseModel):
    """Gemessener Verbrauch eines Runs oder einer Teilmenge (Stage/Provider/Modell).

    Optionale Felder bleiben None, wenn die Datenquelle sie nicht liefert —
    der jeweilige Status macht die Datenqualität explizit.
    """

    model_config = _STRICT

    input_tokens: Optional[int] = Field(None, ge=0)
    output_tokens: Optional[int] = Field(None, ge=0)
    total_tokens: Optional[int] = Field(None, ge=0)
    llm_calls: int = Field(0, ge=0)
    cost_micros: Optional[int] = Field(None, ge=0)
    cost_status: CostStatus = "unknown"
    tokens_status: TokensStatus = "unknown"
    duration_ms: int = Field(0, ge=0)


class RunUsage(BaseModel):
    """Tatsächlicher Verbrauch eines Runs, aufgeschlüsselt nach Stage/Provider/Modell."""

    model_config = _STRICT

    schema_version: Literal[1] = 1
    totals: UsageMetrics
    by_stage: dict[str, UsageMetrics] = Field(default_factory=dict)
    by_provider: dict[str, UsageMetrics] = Field(default_factory=dict)
    by_model: dict[str, UsageMetrics] = Field(default_factory=dict)
    started_at: Optional[str] = None
    ended_at: Optional[str] = None
    measurement_status: MeasurementStatus = "unknown"
    # Version/Quelle der für Kosten verwendeten Providerpreise (Audit, #764)
    pricing_version: Optional[str] = None
    pricing_source: Optional[str] = None


class RunBudgetStatus(BaseModel):
    """Aktueller Budget-Zustand eines Runs: Config + Verbrauch + Warnungen."""

    model_config = _STRICT

    config: RunBudgetConfig
    consumed: UsageMetrics
    warnings: list[BudgetWarning] = Field(default_factory=list)
    status: BudgetState = "ok"
    exceeded_dimension: Optional[BudgetDimension] = None


class PreflightModelRef(BaseModel):
    """Ein für den Run aufgelöstes Modell samt Kostenklarheit."""

    model_config = _STRICT

    stage: str = Field(min_length=1)
    provider_id: str = Field(min_length=1)
    model_id: str = Field(min_length=1)
    base_url_sanitized: Optional[str] = None
    cost_status: CostStatus = "unknown"


class PreflightEstimate(BaseModel):
    """Ehrlich gekennzeichnete Schätzung vor dem Run-Start.

    Liefert Bereiche statt Pseudoexaktheit. Bei unzureichender Datenbasis
    bleiben die jeweiligen Felder None und data_quality/warnings erklären
    warum.
    """

    model_config = _STRICT

    schema_version: Literal[1] = 1
    is_estimate: Literal[True] = True
    estimated_tokens_low: Optional[int] = Field(None, ge=0)
    estimated_tokens_high: Optional[int] = Field(None, ge=0)
    estimated_cost_micros_low: Optional[int] = Field(None, ge=0)
    estimated_cost_micros_high: Optional[int] = Field(None, ge=0)
    estimated_duration_seconds_low: Optional[int] = Field(None, ge=0)
    estimated_duration_seconds_high: Optional[int] = Field(None, ge=0)
    cost_status: CostStatus = "unknown"
    models: list[PreflightModelRef] = Field(default_factory=list)
    pricing_version: str = Field(min_length=1)
    pricing_source: str = Field(min_length=1)
    data_quality: DataQuality = "unknown"
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/contracts/runBudgetContract.ts
/**
 * Run-Budget-Contract v1 — Zod-Spiegel (Issue #764).
 *
 * Hand-gepflegt, 1:1 zu:
 *   schemas/run-budget-config.schema.json
 *   schemas/run-usage.schema.json
 *   schemas/run-budget-status.schema.json
 *   schemas/run-preflight-estimate.schema.json
 *
 * Änderungen am Pydantic-Modell (backend/app/contracts/run_budget_contract.py)
 * → Schema-Dump → diese Datei synchronisieren.
 *
 * Regeln (gespiegelt aus dem Backend-Contract):
 *   - Geldbeträge ausschließlich als Integer-Micros (1 Einheit = 10^-6 Währung).
 *   - Unbekannte Werte sind null + Status ("unknown"/"estimated"/"free"),
 *     niemals 0 für "nicht gemessen".
 */
import { z } from "zod";

// === Literale (1:1 zum Pydantic-Contract) ===
export const BudgetDimensionSchema = z.enum(["tokens", "cost", "time", "calls"]);
export type BudgetDimension = z.infer<typeof BudgetDimensionSchema>;

export const BudgetEnforcementSchema = z.enum(["soft", "hard"]);
export type BudgetEnforcement = z.infer<typeof BudgetEnforcementSchema>;

export const CostStatusSchema = z.enum([
  "measured",
  "estimated",
  "free",
  "unknown",
]);
export type CostStatus = z.infer<typeof CostStatusSchema>;

export const TokensStatusSchema = z.enum(["measured", "partial", "unknown"]);
export type TokensStatus = z.infer<typeof TokensStatusSchema>;

export const MeasurementStatusSchema = z.enum([
  "complete",
  "partial",
  "unknown",
]);
export type MeasurementStatus = z.infer<typeof MeasurementStatusSchema>;

export const DataQualitySchema = z.enum(["high", "medium", "low", "unknown"]);
export type DataQuality = z.infer<typeof DataQualitySchema>;

export const BudgetStateSchema = z.enum(["ok", "warning", "exceeded"]);
export type BudgetState = z.infer<typeof BudgetStateSchema>;

export const TerminationReasonSchema = z.enum([
  "completed",
  "error",
  "user_cancel",
  "user_stop",
  "budget_tokens",
  "budget_cost",
  "budget_time",
  "budget_calls",
]);
export type TerminationReason = z.infer<typeof TerminationReasonSchema>;

// === RunBudgetConfig (schemas/run-budget-config.schema.json) ===
export const RunBudgetConfigSchema = z
  .object({
    schema_version: z.literal(1).default(1),
    max_tokens: z.number().int().min(1).nullable().optional(),
    max_cost_micros: z.number().int().min(1).nullable().optional(),
    max_duration_seconds: z.number().int().min(1).nullable().optional(),
    max_llm_calls: z.number().int().min(1).nullable().optional(),
    enforcement: BudgetEnforcementSchema.default("soft"),
    currency: z.string().length(3).default("USD"),
  })
  .strict();
export type RunBudgetConfig = z.infer<typeof RunBudgetConfigSchema>;

// === BudgetWarning ===
export const BudgetWarningSchema = z
  .object({
    dimension: BudgetDimensionSchema,
    severity: BudgetEnforcementSchema,
    threshold: z.number().int().min(0),
    observed: z.number().int().min(0),
    message: z.string().min(1),
    ts: z.string().min(1),
  })
  .strict();
export type BudgetWarning = z.infer<typeof BudgetWarningSchema>;

// === UsageMetrics ===
export const UsageMetricsSchema = z
  .object({
    input_tokens: z.number().int().min(0).nullable().optional(),
    output_tokens: z.number().int().min(0).nullable().optional(),
    total_tokens: z.number().int().min(0).nullable().optional(),
    llm_calls: z.number().int().min(0).default(0),
    cost_micros: z.number().int().min(0).nullable().optional(),
    cost_status: CostStatusSchema.default("unknown"),
    tokens_status: TokensStatusSchema.default("unknown"),
    duration_ms: z.number().int().min(0).default(0),
  })
  .strict();
export type UsageMetrics = z.infer<typeof UsageMetricsSchema>;

// === RunUsage (schemas/run-usage.schema.json) ===
export const RunUsageSchema = z
  .object({
    schema_version: z.literal(1).default(1),
    totals: UsageMetricsSchema,
    by_stage: z.record(z.string(), UsageMetricsSchema).default(() => ({})),
    by_provider: z.record(z.string(), UsageMetricsSchema).default(() => ({})),
    by_model: z.record(z.string(), UsageMetricsSchema).default(() => ({})),
    started_at: z.string().nullable().optional(),
    ended_at: z.string().nullable().optional(),
    measurement_status: MeasurementStatusSchema.default("unknown"),
    pricing_version: z.string().nullable().optional(),
    pricing_source: z.string().nullable().optional(),
  })
  .strict();
export type RunUsage = z.infer<typeof RunUsageSchema>;

// === RunBudgetStatus (schemas/run-budget-status.schema.json) ===
export const RunBudgetStatusSchema = z
  .object({
    config: RunBudgetConfigSchema,
    consumed: UsageMetricsSchema,
    warnings: z.array(BudgetWarningSchema).default(() => []),
    status: BudgetStateSchema.default("ok"),
    exceeded_dimension: BudgetDimensionSchema.nullable().optional(),
  })
  .strict();
export type RunBudgetStatus = z.infer<typeof RunBudgetStatusSchema>;

// === PreflightModelRef ===
export const PreflightModelRefSchema = z
  .object({
    stage: z.string().min(1),
    provider_id: z.string().min(1),
    model_id: z.string().min(1),
    base_url_sanitized: z.string().nullable().optional(),
    cost_status: CostStatusSchema.default("unknown"),
  })
  .strict();
export type PreflightModelRef = z.infer<typeof PreflightModelRefSchema>;

// === PreflightEstimate (schemas/run-preflight-estimate.schema.json) ===
export const PreflightEstimateSchema = z
  .object({
    schema_version: z.literal(1).default(1),
    is_estimate: z.literal(true),
    estimated_tokens_low: z.number().int().min(0).nullable().optional(),
    estimated_tokens_high: z.number().int().min(0).nullable().optional(),
    estimated_cost_micros_low: z.number().int().min(0).nullable().optional(),
    estimated_cost_micros_high: z.number().int().min(0).nullable().optional(),
    estimated_duration_seconds_low: z
      .number()
      .int()
      .min(0)
      .nullable()
      .optional(),
    estimated_duration_seconds_high: z
      .number()
      .int()
      .min(0)
      .nullable()
      .optional(),
    cost_status: CostStatusSchema.default("unknown"),
    models: z.array(PreflightModelRefSchema).default(() => []),
    pricing_version: z.string().min(1),
    pricing_source: z.string().min(1),
    data_quality: DataQualitySchema.default("unknown"),
    warnings: z.array(z.string()).default(() => []),
  })
  .strict();
export type PreflightEstimate = z.infer<typeof PreflightEstimateSchema>;
